Feature/specify geometry

blairlyons opened the pull request:
Problem
=======
Users can't specify geometry information to be saved in .simularium files
closes #41 

Solution
========
Add a `DisplayData` object to collect geometry information from users. When writing the type mapping for a .simularium file, use this info to create the geometry block within the type mapping.

All the different converters had different ways of collecting per-type information, previously just display names and radii. Sometimes the output files have them and sometimes they don't, sometimes the user could override it and sometimes not. To make things more consistent, I moved name and radius to `DisplayData` and updated all the converters.

Note: this change is breaking imo since a trajectoryInfo version of 3 currently does not load in the viewer.

## Type of change
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] This change requires updated or new tests

Change summary:
---------------
* Added `DisplayData` object to store geometry (display type, url, and color) as well as display names and radii
* When the type mapping is created in `AgentData`, use `DisplayData` to add a geometry section to the type mapping
* Update the converter for each engine so users can use `DisplayData` to specify geometry as well as display names and radii
* Add a check that whenever it is specified, display_type is consistent with viz_type and n_subpoints
* Make default parameters for collections None
* Update tests, docs, and examples

Keyfiles:
-----------------------
1. simulariumio/constants.py - added `DISPLAY_TYPE` with the display options
2. simulariumio/cytosim/cytosim_agent_info.py - removed this object and replaced it with a more general `DisplayData` object that other converters can use
3. simulariumio/cytosim/cytosim_converter.py - update to use new `DisplayData` and set `AgentData`'s `display_data`
4. simulariumio/data_objects/agent_data.py - add `display_data` to hold geometry info, add display data to type mapping when it's generated in `get_type_ids_and_mapping()`, add reading display data in `from_buffer_data()`
5. simulariumio/data_objects/display_data.py - added `DisplayData` object to hold names, radii, and geometry data
6. simulariumio/file_converter.py - update files read in to v3, all the new fields are optional so don't need to add any defaults
7. simulariumio/mcell/mcell_converter.py - update to use new `DisplayData` and set `AgentData.display_data`
8. simulariumio/mcell/mcell_data.py - replace `display_names` with `display_data `
9. simulariumio/medyan/medyan_agent_info.py - removed this object and replaced it with a more general `DisplayData` object that other converters can use
10. simulariumio/medyan/medyan_converter.py - update to use new `DisplayData` and set `AgentData.display_data`
11. simulariumio/medyan/medyan_data.py - replaced `agent_info` with `display_data` to store `DisplayData` for each agent, use `agents_with_endpoints` list to specify agents to draw endpoints for
12. simulariumio/physicell/physicell_converter.py - update to use new `DisplayData` and set `AgentData.display_data`
13. simulariumio/physicell/physicell_data.py - replace `types` mapping with `display_data` to store DisplayData, and `phase_names` to store cell phase names
14. simulariumio/readdy/readdy_converter.py - update to use new `DisplayData` and set `AgentData.display_data`
15. simulariumio/readdy/readdy_data.py - replace `radii` with `display_data`
16. simulariumio/smoldyn/smoldyn_converter.py - update to use new `DisplayData` and set `AgentData.display_data`
17. simulariumio/smoldyn/smoldyn_data.py - replace `radii` and `display_names` with `display_data`
18. simulariumio/springsalad/springsalad_converter.py - update to use new `DisplayData` and set `AgentData.display_data`
19. simulariumio/springsalad/springsalad_data.py - replace 'display_names` with `display_data`
20. simulariumio/trajectory_converter.py - add `_check_types_match_subpoints()` function and check before creating the buffer format, increment trajectory info version to write

meganrm commented on simulariumio/data_objects/display_data.py:
```diff
@@ -0,0 +1,89 @@
+            A float radius for rendering this agent.
+            For fibers, this is the thickness of the line
+            For default agents, this is the scale of the representation
+            Default : 1.0
```
why isn't the default set in `__init__` for this? Is it handled somewhere else?

blairlyons replied:
yeah some of the converters use it as an override only if it has more than a default value, otherwise they use the value in the data. If it isn't specified in the data or here, it does default to 1.0 in the converters. I could set it to 1.0 here but it's harder to do the float comparison than a None check.

blairlyons replied:
added a comment about this

meganrm commented on simulariumio/data_objects/trajectory_data.py:
```diff
@@ -29,9 +29,9 @@ def __init__(
-        plots: List[Dict[str, Any]] = [],
+        time_units: UnitData = None,
+        spatial_units: UnitData = None,
+        plots: List[Dict[str, Any]] = None,
```
is this change unrelated to the display data?

blairlyons replied:
yeah this is part of "make default parameters for collections None" to prevent future bugs, if they default to e.g. `[]` there can be weirdness when using them.

meganrm commented on simulariumio/data_objects/agent_data.py:
```diff
@@ -178,6 +186,13 @@ def get_type_ids_and_mapping(self) -> Tuple[np.ndarray, Dict[str, Any]]:
+                    ):
+                        type_name_mapping[str(tid)]["geometry"] = dict(
+                            self.display_data[tn]
+                        )
```
I don't understand why you're initializing with one default, (SPHERE) but then ignoring the data if only has a default. It seems like you could either have no defaults, and then you just check if the object is empty, or you could just write out default data if none is give so we don't have to to deal with optional data.

blairlyons replied:
yeah, I started out without the "FIBER" display_type and this is leftover from that. I just updated it to default to None and only output the display_type if it's been set to something else.

meganrm commented on simulariumio/tests/converters/test_cytosim_converter.py:
```diff
@@ -322,38 +336,49 @@
             {
                 "trajectoryInfo": {
-                    "version": 2,
+                    "version": 3,
```
it would be good to make a CURRENT_VERSION constant held in the general tests folder that you import to all these files so you don't have to edit each one each time we bump the version

meganrm replied:
you could even put it in here: https://github.com/allen-cell-animated/simulariumio/blob/main/simulariumio/__init__.py

and have a get_file_format_version. Or you could just put it in constants.py. either way, you shouldn't be writing it in every one of these files

blairlyons replied:
yeah added it to constants, there's a separate version for each block

meganrm commented on simulariumio/tests/converters/test_cytosim_converter.py:
```diff
@@ -384,11 +409,26 @@
+                                "displayType": "FIBER",
+                                "color": "#ffc100",
+                            },
+                        },
```
with this whole expected result, it seems like it would be good to have the same expected simularium file for all your tests, so that again, there are less changes you need to make when you update the format.

blairlyons replied:
yeah moved it to conftest, only saves one repeat

meganrm commented on simulariumio/file_converter.py:
```diff
@@ -82,5 +82,8 @@ def update_trajectory_info_version(self, data: Dict[str, Any]) -> Dict[str, Any]
                 "fovDegrees": DEFAULT_CAMERA_SETTINGS.FOV_DEGREES,
             }
             data["trajectoryInfo"]["version"] = 2
```
shouldn't this also now be v = 3?

I think these (heere and line 87) should be `data["trajectoryInfo"]["version"] = self.current_trajectory_info_version`

blairlyons replied:
No, this block updates 1 to 2, then lines 85-87 update 2 to 3. Guess I could combine them but this gives me a place to comment that there are no mandatory fields added, and keeps the structure consistent with transitions that will need defaults added.

meganrm replied:
oh ok, yeah, you should definitely have a block that checks for v == 2, like you say for clarity. I just didn't read this carefully enough and was assuming the other checks were `if else`s. You could still have the separate check for v == 2, but still update v1 ==> v3 data, but it doesn't really matter either way.

meganrm commented on simulariumio/file_converter.py:
```diff
@@ -82,5 +82,8 @@ def update_trajectory_info_version(self, data: Dict[str, Any]) -> Dict[str, Any]
+        if int(data["trajectoryInfo"]["version"]) == 2:
+            # all the new fields from v2 to v3 are optional
+            data["trajectoryInfo"]["version"] = 3
         print(f"Updated TrajectoryInfo v1 -> v{self.current_trajectory_info_version}")
```
this print statement should be in the v ==1 block.

blairlyons replied:
No I just shouldn't hardcode v1, fixed now

toloudis commented on simulariumio/constants.py:
```diff
@@ -45,3 +45,18 @@ class DEFAULT_CAMERA_SETTINGS:
+    GIZMO: str = "GIZMO"
+    PDB: str = "PDB"
+    OBJ: str = "OBJ"
+    FIBER: str = "FIBER"
```
I would suggest removing CUBE and GIZMO at least until we actually support them (or e.g. comment them out and put a comment that says "coming up")

toloudis commented on simulariumio/data_objects/agent_data.py:
```diff
@@ -197,6 +212,27 @@ def get_type_names(
         return result
 
+    @staticmethod
+    def get_display_data(type_mapping: Dict[str, Any]) -> Dict[str, DisplayData]:
```
is there a well defined type for the values in type_mapping?

blairlyons replied:
No, it's coming from a JSON file that was read back in. The keys are strings, but the values are strings, numbers, or another dict, depending on if each agent type has geometry info. I added to the function comment to clarify

toloudis commented on simulariumio/data_objects/display_data.py:
```diff
@@ -0,0 +1,94 @@
+class DisplayData:
+    name: str
+    radius: float
+    display_type: str
```
instead of str, this should be DISPLAY_TYPE?

blairlyons replied:
The fields in DISPLAY_TYPE are strings, so I was thinking if you do `DisplayData(display_type=DISPLAY_TYPE.SPHERE)` the DisplayData object you create has a string display_type = "SPHERE". I'm not as familiar with Python objects, I can make the type DISPLAY_TYPE? seems like then I'd need each display type to be a class deriving from DISPLAY_TYPE?

toloudis replied:
I was picturing something like an Enum : https://docs.python.org/3/library/enum.html
It would be nice to be able to treat DISPLAY_TYPE as a type.  Something to think about anyway...
Also you might be able to do 
`display_type : DISPLAY_TYPE.SPHERE | DISPLAY_TYPE.FIBER...etc`
but 
`display_type: DISPLAY_TYPE` would be nicer

blairlyons replied:
aha yeah I didn't realize I could do enums like that, it's updated now

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -44,6 +42,7 @@ def update_trajectory_info_version(self, data: Dict[str, Any]) -> Dict[str, Any]
         to match the current version
         """
+        original_version = int(data["trajectoryInfo"]["version"])
         if int(data["trajectoryInfo"]["version"]) == 1:
```
line 46 should use `original_version` now

blairlyons replied:
I can set it up differently, but these loops are not if else, so if the version starts at v1, it will get updated to v2 in the first loop, then v2 -> v3 in the next loop, etc. So it needs to keep checking the current version, not the cached original version. 

I'm only storing the original version here so I can report out which version it started from at the end. I changed it to just report that at the beginning of the function to be less confusing. I would get rid of the `original_version` variable completely but I can't embed `int(data["trajectoryInfo"]["version"])` in the f-string. I could also use `original_version` for the first loop logic, but unless I increment it in that loop, I'll need to keep checking the current version in the following `if` loops, not the original version.

toloudis replied:
i see, then my other suggestion is to stash `trajectory_info = data["trajectory_info"]` very early on.  It just gets to be a whole lot of string lookups where you could just do it once and it makes the code a little cleaner.

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -82,5 +81,11 @@ def update_trajectory_info_version(self, data: Dict[str, Any]) -> Dict[str, Any]
             }
             data["trajectoryInfo"]["version"] = 2
-        print(f"Updated TrajectoryInfo v1 -> v{self.current_trajectory_info_version}")
+        if int(data["trajectoryInfo"]["version"]) == 2:
```
is this still `original_version` here?

blairlyons replied:
no because if it started as v1, the data should now be at v2 from going through the first loop and the `original_version` is still 1, no longer current. 

Megan was also confused by how I set up this function, but I didn't want to have to write a loop to go from v1 -> v2 and another to go from v1 -> v4 and have to redundantly redo the fields that are added between v1 and v2 in each loop. Is there a better way to write it though?

meganrm replied:
You could pull the converters out as functions. So the first if block could call, ‘convertv1tov2’ and convertv2tov3’, successively on the data.  And the second if else block would just call ‘convertv2tov3’

toloudis replied:
i like that idea

blairlyons replied:
ok did that

blairlyons commented on simulariumio/data_objects/display_data.py:
```diff
@@ -0,0 +1,94 @@
+        default = self.is_default()
+        return (
+            f"{self.display_type}: url={self.url}, color={self.color} "
+            f"DEFAULT? {default}"
```
I'm returning this string regardless of if_default(), I thought I couldn't embed it in the f-string itself, but I can so I updated it

toloudis replied:
I was just thinking the string `DEFAULT?` didn't look very nice :)

toloudis commented on simulariumio/trajectory_converter.py:
```diff
@@ -358,6 +361,43 @@ def _check_agent_ids_are_unique_per_frame(buffer_data: Dict[str, Any]) -> bool:
         return True
 
+    @staticmethod
+    def _check_types_match_subpoints(trajectory_data: TrajectoryData) -> str:
```
just curious, I wonder if this could be done per-agent rather than iterating over all the data.

blairlyons commented:
@toloudis and @meganrm here's a diff for the changes I made today https://github.com/allen-cell-animated/simulariumio/compare/6ff12e7...feature/specify-geometry, I think I addressed all your comments, lmk if not

toloudis commented on simulariumio/cytosim/cytosim_converter.py:
```diff
@@ -254,6 +255,20 @@ def _read(input_data: CytosimData) -> TrajectoryData:
+                    display_data.display_type = DISPLAY_TYPE.FIBER
+                    print(
+                        f"{display_data.name} display type of "
+                        f"{display_data.display_type} was changed to FIBER"
```
this needs to print the stringified version of the display_type.  (see if the enum can be modified to hold the string values maybe!)

toloudis commented on simulariumio/data_objects/display_data.py:
```diff
@@ -0,0 +1,126 @@
+        self.color = color
+
+    @staticmethod
+    def display_type_to_string(display_type: DISPLAY_TYPE) -> str:
```
i commented on the commit but this could be a member function of the enum class rather than living here.  alternatively you might be able to give the enum values as strings so as to avoid this function completely.  

but more importantly, a comment would be good just saying that these strings are permanent for file i/o reasons and any change to them must incur a version bump on the format.

blairlyons replied:
whoops didn't realize I could use strings as enum values, fixed that and added a comment

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -34,53 +32,78 @@ def __init__(self, input_path: str):
+    @staticmethod
+    def update_trajectory_info_v1_to_v2(data: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Update the trajectory info block from v1 to v2
```
add a docstring comment saying that this mutates the object passed in, and returns it.  it does not return a new copy

toloudis replied:
also, why not pass in data["trajectoryInfo"] instead?  will these functions ever modify any other part of `data`?

toloudis commented on simulariumio/medyan/medyan_converter.py:
```diff
@@ -220,6 +239,20 @@ def _read(input_data: MedyanData) -> TrajectoryData:
+                    display_data.display_type = DISPLAY_TYPE.FIBER
+                    print(
+                        f"{display_data.name} display type of "
+                        f"{display_data.display_type} was changed to FIBER"
```
again, make sure this is printing out the string form of the enum value
